File: packages/codex-thread-bridge/src/codex_thread_bridge/ledger.py
"""Durable request deduplication, including interrupted and partial operations."""

import hashlib
import json
import os
import sqlite3
import time
from pathlib import Path

# The one status a retained receipt can be retried from. It says this request began nothing, so
# repeating it cannot repeat an effect. Every other status keeps the conservative contract, and
# in_progress_or_unknown keeps it for a reason worth stating: what a request began is recorded in
# memory and dies with the process, so a row left behind by a crash cannot say which side of the
# send it stopped on.
RETRYABLE_STATUSES = frozenset({"not_attempted"})

# How many earlier attempts a re-armed receipt keeps: enough to diagnose a flapping socket,
# bounded so a caller that retries all day cannot grow one row without limit.
PRIOR_ATTEMPTS_KEPT = 5


class Ledger:
    def __init__(self, path: Path):
        path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
        descriptor = os.open(path, os.O_CREAT | os.O_RDWR, 0o600)
        os.close(descriptor)
        self.db = sqlite3.connect(path, timeout=10)
        self.db.execute("""
            CREATE TABLE IF NOT EXISTS operations (
                request_id TEXT PRIMARY KEY,
                fingerprint TEXT NOT NULL,
                receipt TEXT NOT NULL
            )
        """)
        self.db.commit()

    @staticmethod
    def _fingerprint(request_id: str, method: str, params: dict):
        if not request_id or len(request_id) > 128:
            raise ValueError("request_id must contain 1–128 characters")
        return hashlib.sha256(
            json.dumps([method, params], sort_keys=True, separators=(",", ":")).encode()
        ).hexdigest()

    def lookup(self, request_id: str, method: str, params: dict, *, legacy_params=None):
        fingerprint = self._fingerprint(request_id, method, params)
        row = self.db.execute(
            "SELECT fingerprint, receipt FROM operations WHERE request_id = ?", (request_id,)
        ).fetchone()
        if row is None:
            return None
        receipt = json.loads(row[1])
        if row[0] != fingerprint:
            legacy_match = (
                receipt.get("fingerprintVersion", 1) == 1
                and legacy_params is not None
                and row[0] == self._fingerprint(request_id, method, legacy_params())
            )
            if not legacy_match:
                raise ValueError(
                    "request_id already belongs to different arguments; no action taken"
                )
        return receipt

    def begin(self, request_id: str, method: str, params: dict, *, legacy_params=None):
        fingerprint = self._fingerprint(request_id, method, params)
        receipt = {
            "requestId": request_id,
            "operation": method,
            "status": "in_progress_or_unknown",
            "startedAt": time.time(),
            "retrySafe": False,
            "fingerprintVersion": 2,
        }
        with self.db:
            inserted = self.db.execute(
                "INSERT OR IGNORE INTO operations VALUES (?, ?, ?)",
                (request_id, fingerprint, json.dumps(receipt)),
            ).rowcount
            existing = self.lookup(request_id, method, params, legacy_params=legacy_params)
            if inserted:
                return True, existing
            if existing.get("status") in RETRYABLE_STATUSES:
                return self._rearm(request_id, receipt, existing)
        return False, existing

    def _rearm(self, request_id: str, receipt: dict, previous: dict):
        """Re-open a request that began nothing, keeping what its earlier attempts reported.

        Two processes sharing a state directory cannot both re-arm one row, and the guard is the
        transaction rather than a comparison: the ignored INSERT above already took this
        transaction's write lock, so the second process waits, then reads the in-progress receipt
        this one wrote and does not re-arm. The fingerprint column is never rewritten, which
        leaves a legacy row's recovery path exactly as it was.
        """
        history = [
            *previous.get("priorAttempts", []),
            {
                "status": previous.get("status"),
                "error": previous.get("error"),
                "updatedAt": previous.get("updatedAt"),
            },
        ][-PRIOR_ATTEMPTS_KEPT:]
        rearmed = {**receipt, "attempt": previous.get("attempt", 1) + 1, "priorAttempts": history}
        self.db.execute(
            "UPDATE operations SET receipt = ? WHERE request_id = ?",
            (json.dumps(rearmed), request_id),
        )
        return True, rearmed

    def save(self, receipt: dict):
        receipt = {**receipt, "updatedAt": time.time()}
        with self.db:
            self.db.execute(
                "UPDATE operations SET receipt = ? WHERE request_id = ?",
                (json.dumps(receipt), receipt["requestId"]),
            )
        return receipt

    def get(self, request_id: str):
        row = self.db.execute(
            "SELECT receipt FROM operations WHERE request_id = ?", (request_id,)
        ).fetchone()
        if row is None:
            raise ValueError("Unknown request_id")
        return json.loads(row[0])

    def close(self):
        self.db.close()

    def import_legacy(self, path: Path):
        """Copy an old alias ledger without losing or silently resolving conflicting receipts."""
        source = sqlite3.connect(path.as_uri() + "?mode=ro", uri=True)
        try:
            rows = source.execute(
                "SELECT request_id, fingerprint, receipt FROM operations"
            ).fetchall()
        finally:
            source.close()
        with self.db:
            for request_id, fingerprint, receipt in rows:
                existing = self.db.execute(
                    "SELECT fingerprint, receipt FROM operations WHERE request_id = ?",
                    (request_id,),
                ).fetchone()
                if existing is not None:
                    if existing[0] != fingerprint or json.loads(existing[1]) != json.loads(receipt):
                        raise ValueError(
                            f"Conflicting retained request {request_id!r} in legacy socket ledger "
                            f"{path}; no requests will be dispatched. Preserve both ledgers."
                        )
                else:
                    self.db.execute(
                        "INSERT INTO operations VALUES (?, ?, ?)",
                        (request_id, fingerprint, receipt),
                    )


def open_endpoint_ledger(socket_path: Path, state_dir: Path):
    """Use one ledger for a canonical endpoint, importing the supplied legacy alias if present."""
    supplied = socket_path.expanduser().absolute()
    canonical = supplied.resolve()
    state_dir = state_dir.expanduser().resolve()

    def filename(path):
        endpoint = hashlib.sha256(str(path).encode()).hexdigest()[:16]
        return state_dir / f"operations-{endpoint}.sqlite3"

    ledger = Ledger(filename(canonical))
    legacy = filename(supplied)
    try:
        if legacy != filename(canonical) and legacy.exists():
            ledger.import_legacy(legacy)
    except BaseException:
        ledger.close()
        raise
    return canonical, ledger

File: internal/bridge/ledger/ledger.go
package ledger

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"
	"unicode/utf8"

	_ "modernc.org/sqlite"
)

const PriorAttemptsKept = 5

//lint:ignore ST1005 ledger.py:125 caller-visible message kept byte-identical to Python
var ErrUnknown = errors.New("Unknown request_id")
var ErrConflict = errors.New("request_id already belongs to different arguments; no action taken")

type Ledger struct{ db *sql.DB }
type Receipt map[string]any

func Open(path string) (*Ledger, error) {
	if err := os.MkdirAll(filepath.Dir(path), 0700); err != nil {
		return nil, fmt.Errorf("ledger directory: %w", err)
	}
	file, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR, 0600)
	if err != nil {
		return nil, fmt.Errorf("ledger file: %w", err)
	}
	if err := file.Close(); err != nil {
		return nil, fmt.Errorf("ledger file close: %w", err)
	}
	db, err := sql.Open("sqlite", path)
	if err != nil {
		return nil, fmt.Errorf("open ledger: %w", err)
	}
	db.SetMaxOpenConns(1)
	if _, err := db.Exec(`PRAGMA busy_timeout=10000; CREATE TABLE IF NOT EXISTS operations (request_id TEXT PRIMARY KEY, fingerprint TEXT NOT NULL, receipt TEXT NOT NULL)`); err != nil {
		db.Close()
		return nil, fmt.Errorf("create ledger: %w", err)
	}
	return &Ledger{db}, nil
}
func (l *Ledger) Close() error { return l.db.Close() }

func Endpoint(socket, state string) (string, *Ledger, error) {
	supplied, err := filepath.Abs(socket)
	if err != nil {
		return "", nil, fmt.Errorf("absolute socket: %w", err)
	}
	canonical, err := filepath.EvalSymlinks(supplied)
	if errors.Is(err, os.ErrNotExist) {
		// Resolve dangling aliases too: Path.resolve(strict=False) follows the
		// symlink before accepting its missing final target.
		current := supplied
		for range 40 {
			parent, parentErr := filepath.EvalSymlinks(filepath.Dir(current))
			if parentErr != nil {
				return "", nil, fmt.Errorf("canonical socket parent: %w", parentErr)
			}
			current = filepath.Join(parent, filepath.Base(current))
			target, linkErr := os.Readlink(current)
			if errors.Is(linkErr, os.ErrNotExist) || errors.Is(linkErr, os.ErrInvalid) {
				canonical = current
				break
			}
			if linkErr != nil {
				return "", nil, fmt.Errorf("canonical socket link: %w", linkErr)
			}
			if filepath.IsAbs(target) {
				current = target
			} else {
				current = filepath.Join(parent, target)
			}
		}
		if canonical == "" {
			return "", nil, fmt.Errorf("canonical socket: too many symlinks")
		}
	} else if err != nil {
		return "", nil, fmt.Errorf("canonical socket: %w", err)
	}
	state, err = filepath.Abs(state)
	if err != nil {
		return "", nil, fmt.Errorf("absolute state: %w", err)
	}
	name := func(path string) string {
		hash := sha256.Sum256([]byte(path))
		return filepath.Join(state, "operations-"+hex.EncodeToString(hash[:])[:16]+".sqlite3")
	}
	l, err := Open(name(canonical))
	if err != nil {
		return "", nil, err
	}
	if canonical != supplied {
		if _, err := os.Stat(name(supplied)); err == nil {
			if err := l.ImportLegacy(context.Background(), name(supplied)); err != nil {
				l.Close()
				return "", nil, err
			}
		} else if !errors.Is(err, os.ErrNotExist) {
			l.Close()
			return "", nil, fmt.Errorf("legacy ledger: %w", err)
		}
	}
	return canonical, l, nil
}

// checkRequestID is ledger.py _fingerprint's bound: 1-128 characters, counted as Python len.
func checkRequestID(id string) error {
	if n := utf8.RuneCountInString(id); n == 0 || n > 128 {
		return errors.New("request_id must contain 1\u2013128 characters")
	}
	return nil
}

func (l *Ledger) Lookup(ctx context.Context, id, method string, params map[string]any, legacy func() map[string]any) (Receipt, error) {
	if err := checkRequestID(id); err != nil {
		return nil, err
	}
	fingerprint, err := Fingerprint(method, params)
	if err != nil {
		return nil, err
	}
	var stored, raw string
	err = l.db.QueryRowContext(ctx, `SELECT fingerprint, receipt FROM operations WHERE request_id=?`, id).Scan(&stored, &raw)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, fmt.Errorf("lookup: %w", err)
	}
	receipt, err := decode(raw)
	if err != nil {
		return nil, err
	}
	if stored != fingerprint {
		version := receipt["fingerprintVersion"]
		if version == nil {
			version = json.Number("1")
		}
		if version != json.Number("1") || legacy == nil {
			return nil, ErrConflict
		}
		old, err := Fingerprint(method, legacy())
		if err != nil {
			return nil, err
		}
		if old != stored {
			return nil, ErrConflict
		}
	}
	return receipt, nil
}

func (l *Ledger) Begin(ctx context.Context, id, method string, params map[string]any, legacy func() map[string]any) (bool, Receipt, error) {
	if err := checkRequestID(id); err != nil {
		return false, nil, err
	}
	fp, err := Fingerprint(method, params)
	if err != nil {
		return false, nil, err
	}
	receipt := Receipt{"requestId": id, "operation": method, "status": "in_progress_or_unknown", "startedAt": float64(time.Now().UnixNano()) / 1e9, "retrySafe": false, "fingerprintVersion": 2}
	raw, err := json.Marshal(receipt)
	if err != nil {
		return false, nil, err
	}
	tx, err := l.db.BeginTx(ctx, nil)
	if err != nil {
		return false, nil, fmt.Errorf("begin transaction: %w", err)
	}
	defer tx.Rollback()
	result, err := tx.ExecContext(ctx, `INSERT OR IGNORE INTO operations VALUES (?,?,?)`, id, fp, raw)
	if err != nil {
		return false, nil, fmt.Errorf("begin receipt: %w", err)
	}
	inserted, err := result.RowsAffected()
	if err != nil {
		return false, nil, err
	}
	var stored, previousRaw string
	if err := tx.QueryRowContext(ctx, `SELECT fingerprint,receipt FROM operations WHERE request_id=?`, id).Scan(&stored, &previousRaw); err != nil {
		return false, nil, err
	}
	previous, err := decode(previousRaw)
	if err != nil {
		return false, nil, err
	}
	if stored != fp {
		version := previous["fingerprintVersion"]
		if version == nil {
			version = json.Number("1")
		}
		if version != json.Number("1") || legacy == nil {
			return false, nil, ErrConflict
		}
		old, err := Fingerprint(method, legacy())
		if err != nil {
			return false, nil, err
		}
		if old != stored {
			return false, nil, ErrConflict
		}
	}
	if inserted == 0 && previous["status"] == "not_attempted" {
		history, _ := previous["priorAttempts"].([]any)
		history = append(history, Receipt{"status": previous["status"], "error": previous["error"], "updatedAt": previous["updatedAt"]})
		if len(history) > PriorAttemptsKept {
			history = history[len(history)-PriorAttemptsKept:]
		}
		attempt := 1
		if n, ok := previous["attempt"].(json.Number); ok {
			parsed, err := n.Int64()
			if err != nil {
				return false, nil, err
			}
			attempt = int(parsed)
		}
		receipt["attempt"] = attempt + 1
		receipt["priorAttempts"] = history
		raw, err = json.Marshal(receipt)
		if err != nil {
			return false, nil, err
		}
		if _, err = tx.ExecContext(ctx, `UPDATE operations SET receipt=? WHERE request_id=?`, raw, id); err != nil {
			return false, nil, err
		}
		previous = receipt
		inserted = 1
	}
	if err := tx.Commit(); err != nil {
		return false, nil, fmt.Errorf("commit receipt: %w", err)
	}
	return inserted == 1, previous, nil
}
